### PR TITLE
ENH: stats: add ecdf function

### LINKED ISSUE #17431: ENH: ECDF in scipy.
### Is your feature request related to a problem? Please describe.

When in Python, calculating an ECDF requires loading `statsmodels`' [`ECDF`](https://www.statsmodels.org/dev/generated/statsmodels.distributions.empirical_distribution.ECDF.html) function. It is a really big dependency for something relatively small.

### Describe the solution you'd like.

A function to calculate the empirical distribution function of a given sample.

### Describe alternatives you've considered.

Using `statsmodels`.

### Additional context (e.g. screenshots, GIFs)

It was mentioned as a `numpy` feature requrest at some point but they decided against it (reasonably) as it is more of a `scipy` feature: https://github.com/numpy/numpy/issues/9057.

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Closes gh-17431

#### What does this implement/fix?
This PR adds an ECDF (empirical cumulative distribution function) function as requested in gh-17431.

#### Additional information
Follow-up work is planned, but I thought I'd start with the minimum complete function (i.e. uncensored data only, no confidence intervals, no bells and whistles). 

One question is whether  we should follow Matlab and duplicate the minimum value of `sample`. For example, should the docstring example be like:
```
    >>> res = stats.ecdf(sample)
    >>> res.x
    array([5.2 , 5.58, 6.23, 6.42, 7.06])
    >>> res.cdf
    array([0.2, 0.4, 0.6, 0.8, 1. ])
```
or
```
    >>> res.x
    array([5.2, 5.2 , 5.58, 6.23, 6.42, 7.06])  # 5.2 is duplicated
    >>> res.cdf
    array([0, 0.2, 0.4, 0.6, 0.8, 1. ])  # 5.2 also corresponds with CDF of zero
```
My preference is not to do this because equally valid to prepend any point lower than the minimum of the sample with a corresponding CDF of zero. Likewise, any point after the maximum could be appended with a corresponding CDF of 1. The choice is application-dependent. We can add a simple plotting function for convenience later, but we don't need to return the extra points we chose just to make the plot look nice. We could add parameters to control this behavior, but that can be done in a follow-up, too.

@hadjipantelis

### CODE REVIEW

File: scipy/stats/_survival.py
Comment (tupui) [inline scipy/stats/_survival.py:9]: Can we instead declare classes. The issue with doing it like that is that IDEs (also iPython) are not smart enough to follow through and provide autocomplete.

File: scipy/stats/_survival.py
Comment (tupui) [inline scipy/stats/_survival.py:13]:
```suggestion
    """Empirical cumulative distribution function of a sample.
```

File: scipy/stats/_survival.py
Comment (tupui) [inline scipy/stats/_survival.py:25]: What about just keeping this out for this PR? I expect we will have enough time to do the `stats.CensoredData` before the next release.

File: scipy/stats/_survival.py
Comment (tupui) [inline scipy/stats/_survival.py:36]: Is this really needed if in the end we don't do more than `1-CDF`?

File: scipy/stats/_survival.py
Comment (tupui) [inline scipy/stats/_survival.py:61]:
```suggestion
    As in the example from [1]_ page 79, five boys were selected at random from
```

File: scipy/stats/_survival.py
Comment (tupui) [inline scipy/stats/_survival.py:90]:
```suggestion
    """
    if not isinstance(sample, CensoredData):
```

File: scipy/stats/tests/test_survival.py
Comment (tupui) [inline scipy/stats/tests/test_survival.py:12]: Some folks might be tempted to pass n-D samples. I think we should clarify in the doc and also specifically test that.

Looking at the error message, it's also not clear that we want more than 1 obs.

**Review by tupui — COMMENTED** [review]
Thanks @mdhaber, +1 for me. This is really something we need to do all the time and always wanted to have it in here.

I would go for not duplicating info. (The current implementation.)

File: scipy/stats/tests/test_survival.py
Comment (tupui) [inline scipy/stats/tests/test_survival.py:35]: One thing I was not sure here. Because we use `unique` we won't have the same number of output values as the input. I am fine with that since we return both `x` and `cdf`

**Comment by lorentzenchr** [comment]
> My preference is not to do this [duplication of the lowest value]

+1.

Concerning the API, I would have expected to get a full `rv_continuous` (or `rv_discrete`). Then, with your above values
```python
x = [5.2 , 5.58, 6.23, 6.42, 7.06]
my_ecdf = ecdf(x)
x.cdf(0)  # returns 0
x.cdf(5.2)  # returns 0.2
x.cdf(10)  # returns 1
```

**Comment by mdhaber** [comment]
@lorentzenchr I considered that and that is why I waited so long to write this. But what is the pdf/pmf? This is discussed in detail in https://github.com/numpy/numpy/issues/9057#issuecomment-299579238 and below. I don't think there is a great answer, so I decided against this.

**Comment by tupui** [comment]
> @lorentzenchr I considered that, and that is why I waited so long to write this. But what is the pdf/pmf? This is discussed in detail in [numpy/numpy#9057 (comment)](https://github.com/numpy/numpy/issues/9057#issuecomment-299579238) and below. I don't think there is a great answer, so I decided against this.

Running KDE on top?

**Comment by mdhaber** [comment]
KDE is an alternative to ECDF, basically. ECDF is step-wise, so KDE is not consistent with that.

File: scipy/stats/tests/test_survival.py
Comment (mdhaber) [reply scipy/stats/tests/test_survival.py:35]: I don't understand. `ecdf` also takes the unique values.  It returns only the unique values in `sample`, and it returns a CDF associated with each.

File: scipy/stats/_survival.py
Comment (mdhaber) [reply scipy/stats/_survival.py:36]: When we add support for CensoredData, many people will want to use this function to get the SF. ("survival analysis")

File: scipy/stats/tests/test_survival.py
Comment (tupui) [reply scipy/stats/tests/test_survival.py:35]: You could have had this
```
xs = np.sort(data)
ys = np.linspace(0, 1, num=len(data))
```

File: scipy/stats/_survival.py
Comment (mdhaber) [reply scipy/stats/_survival.py:25]: Yes. So wouldn't in or out be OK?

Comment (tupui) [reply scipy/stats/_survival.py:25]: Fine with both, mostly saying this so we don't discuss it more than that during this review.

File: scipy/stats/_survival.py
Comment (tupui) [reply scipy/stats/_survival.py:36]: Ok so conveniency. Fine with me 👍

**Comment by lorentzenchr** [comment]
@mdhaber Thanks for pointing at the discussions.

> But what is the pdf/pmf?

With `rv_discrete` the values of the pmf are well defined. But an empirical `cdf` function is for sure a good start and covers 95% of use cases.

**Comment by hadjipantelis** [comment]
@mdhaber Thank you for this! I think it is reasonable not to duplicate the minimum especially given we return `x` too; ECDF will be a very helpful addition.

**Comment by josef-pkt** [comment]
statsmodels ECDF has additionally a `__call__` method to get cdf value for non-observation points, i.e. full step function

```
    def __call__(self, time):

        tind = np.searchsorted(self.x, time, self.side) - 1
        return self.y[tind]
```

**Comment by mdhaber** [comment]
> With rv_discrete the values of the pmf are well defined

I've considered that, too, but there are two reasons we don't just return an `rv_discrete`:
- `rv_discrete` is broken and would not be a good object to return in its current state. See gh-17528 and the history (gh-17236, gh-8057). The plan of [this project](https://chanzuckerberg.com/eoss/proposals/scipy-fundamental-tools-for-biomedical-research/) is to address this (and the rest of gh-15928) starting this summer.
- the ECDF (and empirical survival function) is often thought of as an approximation of an underlying continuous distribution, so it is not always appropriate to think of it as a discrete distribution.

> statsmodels ECDF has additionally a __call__ method

The object that is returned can certainly be extended to include a `__call__` method, properties, and/or other methods. I prefer to work incrementally; PRs tend to stall due to reviewer fatigue when they get too long. As in the top post, 

> Follow-up work is planned, but I thought I'd start with the minimum complete function (i.e. uncensored data only, no confidence intervals, no bells and whistles).

**Comment by tupui** [comment]
Yes, let's keep this contained and simple.

**Comment by alan-isaac** [comment]
Hoping this implementation will be very collaborative with [statsmodels](https://www.statsmodels.org/dev/generated/statsmodels.distributions.empirical_distribution.ECDF.html).

File: scipy/stats/_survival.py
Comment (tupui) [inline scipy/stats/_survival.py:83]: I think it should be this instead. Adding 4 in the position 0.

```suggestion
    >>> ax.step(np.insert(res.x, 0, 4), np.insert(res.cdf, 0, 0), where='post')
```

File: scipy/stats/_survival.py
Comment (tupui) [inline scipy/stats/_survival.py:48]:
```suggestion
    [1]_; other forms of censoring are not supported at this time.
```

**Review by tupui — APPROVED** [review]
LGTM, once my last comments are addressed I will merge.

**Comment by tupui** [comment]
Just going ahead with the suggestion and merging. As you are working on a follow up either way, it does not really matter.

**Pull request merged**
